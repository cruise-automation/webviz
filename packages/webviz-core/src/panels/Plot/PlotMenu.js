// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import cx from "classnames";
import React from "react";

import styles from "./PlotMenu.module.scss";
import { PanelToolbarLabel, PanelToolbarInput } from "webviz-core/shared/panelToolbarStyles";
import Dropdown from "webviz-core/src/components/Dropdown";
import Item from "webviz-core/src/components/Menu/Item";
import type { PlotConfig } from "webviz-core/src/panels/Plot";
import { type DataSet } from "webviz-core/src/panels/Plot/PlotChart";
import { downloadFiles } from "webviz-core/src/util";
import { formatTimeRaw } from "webviz-core/src/util/time";

function isValidInput(value: string) {
  return value === "" || !isNaN(parseFloat(value));
}

export function getHeader(message: any) {
  let header = null;
  for (const key in message) {
    if (key.includes("header")) {
      header = message[key];
    }
  }
  return header;
}

function formatData(data: any, label: string) {
  const { x, y } = data;
  const { receiveTime, message } = data.tooltip.item.message;
  const receiveTimeFloat = formatTimeRaw(receiveTime);
  const header = getHeader(message);
  const stampTime = header ? formatTimeRaw(header.stamp) : "";
  return [x, receiveTimeFloat, stampTime, label, y];
}

export function getCSVData(datasets: DataSet[]): string {
  const headLine = ["elapsed time", "receive time", "header.stamp", "topic", "value"];
  const combinedLines = [];
  combinedLines.push(headLine);
  datasets.forEach((dataset, idx) => {
    dataset.data.forEach((data) => {
      combinedLines.push(formatData(data, dataset.label));
    });
  });
  return combinedLines.join("\n");
}

function downloadCsvFile(datasets: DataSet[]) {
  const csv = getCSVData(datasets);
  const blob = new Blob([csv], { type: "text/csv;charset=utf-8;" });
  downloadFiles([{ blob, fileName: `plot_data_export.csv` }]);
}

export default function PlotMenu({
  minYValue,
  maxYValue,
  saveConfig,
  setMinMax,
  datasets,
  xAxisVal = "timestamp",
}: {
  minYValue: string,
  maxYValue: string,
  saveConfig: ($Shape<PlotConfig>) => void,
  setMinMax: ($Shape<PlotConfig>) => void,
  datasets: DataSet[],
  xAxisVal?: "timestamp" | "index",
}) {
  return (
    <>
      <Item>
        <PanelToolbarLabel>X-axis</PanelToolbarLabel>
        <Dropdown value={xAxisVal} onChange={(newXAxisVal) => saveConfig({ xAxisVal: newXAxisVal })}>
          <span value="timestamp">timestamp</span>
          <span value="index">index</span>
        </Dropdown>
      </Item>
      <Item onClick={() => saveConfig({ maxYValue: maxYValue === "" ? "10" : "" })}>
        <div className={styles.label}>Maximum</div>
        <PanelToolbarInput
          className={cx(styles.input, { [styles.inputError]: !isValidInput(maxYValue) })}
          value={maxYValue}
          onChange={(event) => {
            saveConfig({ maxYValue: event.target.value });
          }}
          onClick={(event) => event.stopPropagation()}
          placeholder="auto"
        />
      </Item>
      <Item onClick={() => saveConfig({ minYValue: minYValue === "" ? "-10" : "" })}>
        <div className={styles.label}>Minimum</div>
        <PanelToolbarInput
          className={cx(styles.input, { [styles.inputError]: !isValidInput(minYValue) })}
          value={minYValue}
          onChange={(event) => {
            saveConfig({ minYValue: event.target.value });
          }}
          onClick={(event) => event.stopPropagation()}
          placeholder="auto"
        />
      </Item>
      <Item>
        <button onClick={setMinMax}>Set min and max</button>
      </Item>
      <Item>
        <button onClick={() => downloadCsvFile(datasets)}>Download plot data (csv)</button>
      </Item>
    </>
  );
}
