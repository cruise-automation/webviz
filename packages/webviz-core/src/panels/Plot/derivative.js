// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { type TimeBasedChartTooltipData } from "webviz-core/src/components/TimeBasedChart";
import type { PlotChartPoint } from "webviz-core/src/panels/Plot/PlotChart";

export default function derivative(
  data: PlotChartPoint[],
  tooltips: TimeBasedChartTooltipData[],
  includeTooltipInData: boolean
): { points: PlotChartPoint[], tooltips: TimeBasedChartTooltipData[] } {
  const points = [];
  const newTooltips = [];
  for (let i = 1; i < data.length; i++) {
    const secondsDifference = data[i].x - data[i - 1].x;
    const value = (data[i].y - data[i - 1].y) / secondsDifference;
    const previousTooltip = tooltips[i];
    const point: PlotChartPoint = { x: data[i].x, y: value };
    const tooltip = {
      x: point.x,
      y: point.y,
      item: previousTooltip.item,
      path: `${previousTooltip.path}.@derivative`,
      datasetKey: previousTooltip.datasetKey,
      datasetIndex: i - 1,
      value,
      constantName: undefined,
      startTime: previousTooltip.startTime,
    };
    newTooltips.push(tooltip);
    if (includeTooltipInData) {
      point.tooltip = tooltip;
    }
    points.push(point);
  }
  return { points, tooltips: newTooltips };
}
