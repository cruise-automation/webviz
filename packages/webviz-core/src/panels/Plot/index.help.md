# Plot

Plots arbitrary values from topics, similar to [rqt_plot](http://wiki.ros.org/rqt_plot).

The values plotted are specified through Webviz's [message path syntax](/help/message-path-syntax).

In the options menu, you can set a minimum and maximum Y-value. By default, the plot will use those bounds unless the data you're looking at extends above the max Y-value or below the min Y-value; in those cases it will automatically expand. If you want to disable this behavior and force the plot to use the exact minimum and maximum Y-values entered, you can "lock" the y-axis in the options menu as well.

There are 2 possible values for the x-axis - by the message's timestamp or by the message values' respective array indices.

## Timestamp as X-Axis

When you select to plot against a timestamp x-axis (default), you can specify whether the timestamp is taken from either the time the message was received, or the `message.header.stamp` field of the topic, in the dropdown displayed next to each line topic. All Plot panels' timestamped x-axes are kept in sync for easy comparison.

When you select multiple values (by using a slice) the plot turns into a scatter plot at each timestamp tick.

You can also enter an arbitrary number, which will add a horizontal line at that y-value.

To take the derivative of a value (change per second), use the special `.@derivative` modifier. This does not work with scatter plots (when using slices).

To plot the length of an array, use the `.@length` modifier at the end of the queried array.

To switch the sign of a value, use the special `.@negative` modifier at the end of the topic path syntax. The following math functions are also available: `.@abs`, `.@acos`, `.@asin`, `.@atan`, `.@ceil`, `.@cos`, `.@log`, `.@log1p`, `.@log2`, `.@log10`, `.@round`, `.@sign`, `.@sin`, `.@sqrt`, `.@tan`, and `.@trunc`. See the [Javascript Math documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math) for details on each one.

## Array Index as X-Axis

In the legend, you can change the x-axis from being the messages' timestamps to their values' array indices. In this mode, adding message path `/some_topic.some_array` as a new line in the plot will chart that array's values against their indices. For example, if `/some_topic.some_array` contained the values `[5, 10, 15]`, the resulting points on the chart would be `[0, 5]`, `[1, 10]`, and `[2, 15]`.

This mode always plots just the data from the latest tick.

## Message Data as X-Axis

In the legend, set the x-axis mode to a "msg path" option to specify a path to the data to use for the plot's x-coordinates. For example, you could plot a line for `/some_topic.y`, with the x axis set to `/some_topic.x`.

To plot data from just the most recent tick, use the "msg path (current)" setting. To plot data from _all_ matching messages, use the "msg path (accumulated)" setting.

## User Interactions

You can zoom by scrolling, and pan by dragging. Double-click to reset.

By default, scrolling will zoom horizontally only. To zoom vertically, hold the `v` key while scrolling. To zoom both horizontally and vertically at the same time, hold the `b` key while scrolling.

Some "default zoom level" settings are available in the panel settings menu. The y-axis limits can be given fixed values for any plot type. For plots where the x-axis represents the message timestamp, you can have the plot "follow" playback by specifying the plot's "viewport" (i.e. width in seconds).
