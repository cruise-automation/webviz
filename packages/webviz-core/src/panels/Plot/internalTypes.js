// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import type { TimestampMethod } from "webviz-core/src/util/time";

export type BasePlotPath = {
  value: string,
  enabled: boolean,
};

export type PlotPath = BasePlotPath & {
  timestampMethod: TimestampMethod,
};

// A "reference line" plot path is a numeric value. It creates a horizontal line on the plot at the specified value.
export function isReferenceLinePlotPathType(path: BasePlotPath): boolean {
  return !isNaN(Number.parseFloat(path.value));
}
