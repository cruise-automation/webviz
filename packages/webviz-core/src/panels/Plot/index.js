// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { compact, uniq } from "lodash";
import memoizeWeak from "memoize-weak";
import React, { useEffect, useCallback, useMemo, useRef } from "react";
import { hot } from "react-hot-loader/root";
import { type Time, TimeUtil } from "rosbag";

import helpContent from "./index.help.md";
import Flex from "webviz-core/src/components/Flex";
import { type MessageHistoryItemsByPath } from "webviz-core/src/components/MessageHistoryDEPRECATED";
import { getTopicsFromPaths } from "webviz-core/src/components/MessagePathSyntax/parseRosPath";
import { useDecodeMessagePathsForMessagesByTopic } from "webviz-core/src/components/MessagePathSyntax/useCachedGetMessagePathDataItems";
import { useMessagePipeline } from "webviz-core/src/components/MessagePipeline";
import Panel from "webviz-core/src/components/Panel";
import PanelToolbar from "webviz-core/src/components/PanelToolbar";
import { getTooltipItemForMessageHistoryItem, type TooltipItem } from "webviz-core/src/components/TimeBasedChart/utils";
import { useBlocksByTopic, useDataSourceInfo, useMessagesByTopic } from "webviz-core/src/PanelAPI";
import type { BasePlotPath, PlotPath } from "webviz-core/src/panels/Plot/internalTypes";
import PlotChart, { getDatasetsAndTooltips, type PlotDataByPath } from "webviz-core/src/panels/Plot/PlotChart";
import PlotLegend from "webviz-core/src/panels/Plot/PlotLegend";
import PlotMenu from "webviz-core/src/panels/Plot/PlotMenu";
import type { PanelConfig } from "webviz-core/src/types/panels";
import { useShallowMemo, useGetCurrentValue } from "webviz-core/src/util/hooks";
import { fromSec, subtractTimes, toSec } from "webviz-core/src/util/time";

export const plotableRosTypes = [
  "bool",
  "int8",
  "uint8",
  "int16",
  "uint16",
  "int32",
  "uint32",
  "int64",
  "uint64",
  "float32",
  "float64",
  "time",
  "duration",
  "string",
  "json",
  // We plot the length of arrays.
  "array",
];

// X-axis values:
export type PlotXAxisVal =
  | "timestamp" // Message playback time. Preloaded.
  | "index" // Message-path value index. One "current" message at playback time.
  | "custom" // Message path data. Preloaded.
  | "currentCustom"; // Message path data. One "current" message at playback time.

export type PlotConfig = {
  paths: PlotPath[],
  minYValue: string,
  maxYValue: string,
  showLegend: boolean,
  xAxisVal: PlotXAxisVal,
  xAxisPath?: BasePlotPath,
  followingViewWidth?: string,
};

export function openSiblingPlotPanel(
  openSiblingPanel: (string, cb: (PanelConfig) => PanelConfig) => void,
  topicName: string
) {
  openSiblingPanel(
    "Plot",
    (config: PlotConfig) =>
      ({
        ...config,
        paths: uniq(
          config.paths
            .concat([{ value: topicName, enabled: true, timestampMethod: "receiveTime" }])
            .filter(({ value }) => value)
        ),
      }: PlotConfig)
  );
}

type Props = {
  config: PlotConfig,
  saveConfig: ($Shape<PlotConfig>) => void,
};

// messagePathItems contains the whole parsed message, and we don't need to cache all of that.
// Instead, throw away everything but what we need (the timestamps).
const getPlotDataByPath = (itemsByPath: MessageHistoryItemsByPath): PlotDataByPath => {
  const ret: PlotDataByPath = {};
  Object.keys(itemsByPath).forEach((path) => {
    ret[path] = [itemsByPath[path].map(getTooltipItemForMessageHistoryItem)];
  });
  return ret;
};

const getMessagePathItemsForBlock = memoizeWeak(
  (decodeMessagePathsForMessagesByTopic, block): PlotDataByPath => {
    return Object.freeze(getPlotDataByPath(decodeMessagePathsForMessagesByTopic(block)));
  }
);

const ZERO_TIME = { sec: 0, nsec: 0 };

function getBlockItemsByPath(decodeMessagePathsForMessagesByTopic, blocks) {
  const ret = {};
  const lastBlockIndexForPath = {};
  blocks.forEach((block, i) => {
    const messagePathItemsForBlock: PlotDataByPath = getMessagePathItemsForBlock(
      decodeMessagePathsForMessagesByTopic,
      block
    );
    Object.keys(messagePathItemsForBlock).forEach((path) => {
      const existingItems: TooltipItem[][] = ret[path] || [];
      // getMessagePathItemsForBlock returns an array of exactly one range of items.
      const [pathItems] = messagePathItemsForBlock[path];
      if (lastBlockIndexForPath[path] === i - 1) {
        // If we are continuing directly from the previous block index (i - 1) then add to the
        // existing range, otherwise start a new range
        const currentRange = existingItems[existingItems.length - 1];
        for (const item of pathItems) {
          currentRange.push(item);
        }
      } else {
        // Start a new contiguous range. Make a copy so we can extend it.
        existingItems.push(pathItems.slice());
      }
      ret[path] = existingItems;
      lastBlockIndexForPath[path] = i;
    });
  });
  return ret;
}

function Plot(props: Props) {
  const { saveConfig, config } = props;
  const { followingViewWidth, paths: yAxisPaths, minYValue, maxYValue, showLegend, xAxisVal, xAxisPath } = config;
  // Note that the below values are refs since they are only used in callbacks and are not rendered anywhere.
  const currentMinY = useRef(null);
  const currentMaxY = useRef(null);
  const currentViewWidth = useRef(null);

  const saveCurrentView = useCallback((minY: number, maxY: number, width: ?number) => {
    currentMinY.current = minY;
    currentMaxY.current = maxY;
    currentViewWidth.current = width;
  }, []);

  const setWidth = useCallback(
    () =>
      saveConfig({
        followingViewWidth: currentViewWidth.current != null ? currentViewWidth.current.toString() : "",
      }),
    [saveConfig]
  );

  const setMinMax = useCallback(
    () =>
      saveConfig({
        minYValue: currentMinY.current != null ? currentMinY.current.toString() : "",
        maxYValue: currentMaxY.current != null ? currentMaxY.current.toString() : "",
      }),
    [currentMaxY, currentMinY, saveConfig]
  );

  useEffect(() => {
    if (!yAxisPaths.length) {
      saveConfig({ paths: [{ value: "", enabled: true, timestampMethod: "receiveTime" }] });
    }
  });

  const showSingleCurrentMessage = xAxisVal === "currentCustom" || xAxisVal === "index";
  const historySize = showSingleCurrentMessage ? 1 : Infinity;

  const allPaths = yAxisPaths.map(({ value }) => value).concat(compact([xAxisPath?.value]));
  const memoizedPaths: string[] = useShallowMemo<string[]>(allPaths);
  const subscribeTopics = useMemo(() => getTopicsFromPaths(memoizedPaths), [memoizedPaths]);
  const messagesByTopic = useMessagesByTopic({
    topics: subscribeTopics,
    historySize,
    // This subscription is used for two purposes:
    //  1. A fallback for preloading when blocks are not available (nodes, websocket.)
    //  2. Playback-synced plotting of index/custom data.
    preloadingFallback: !showSingleCurrentMessage,
    format: "bobjects",
  });

  const decodeMessagePathsForMessagesByTopic = useDecodeMessagePathsForMessagesByTopic(memoizedPaths);

  const streamedItemsByPath = useMemo(() => getPlotDataByPath(decodeMessagePathsForMessagesByTopic(messagesByTopic)), [
    decodeMessagePathsForMessagesByTopic,
    messagesByTopic,
  ]);

  const blocks = useBlocksByTopic(subscribeTopics);
  const blockItemsByPath = useMemo(
    () => (showSingleCurrentMessage ? {} : getBlockItemsByPath(decodeMessagePathsForMessagesByTopic, blocks)),
    [showSingleCurrentMessage, decodeMessagePathsForMessagesByTopic, blocks]
  );
  const { startTime } = useDataSourceInfo();

  // If every streaming key is in the blocks, just use the blocks object for a stable identity.
  const mergedItems = useMemo(
    () =>
      Object.keys(streamedItemsByPath).every((path) => blockItemsByPath[path] != null)
        ? blockItemsByPath
        : { ...streamedItemsByPath, ...blockItemsByPath },
    [streamedItemsByPath, blockItemsByPath]
  );

  // Don't filter out disabled paths when passing into getDatasetsAndTooltips, because we still want
  // easy access to the history when turning the disabled paths back on.
  const { datasets, tooltips, pathsWithMismatchedDataLengths } = useMemo(
    // TODO(steel): This memoization isn't quite ideal: getDatasetsAndTooltips is a bit expensive
    // with lots of preloaded data, and when we preload a new block we re-generate the datasets for
    // the whole timeline. We should try to use block memoization here.
    () => getDatasetsAndTooltips(yAxisPaths, mergedItems, startTime || ZERO_TIME, xAxisVal, xAxisPath),
    [yAxisPaths, mergedItems, startTime, xAxisVal, xAxisPath]
  );

  const { currentTime, endTime, seekPlayback: seek } = useMessagePipeline(
    useCallback(
      ({ seekPlayback, playerState: { activeData } }) => ({
        currentTime: activeData?.currentTime,
        endTime: activeData?.endTime,
        seekPlayback,
      }),
      []
    )
  );
  // Min/max x-values and playback position indicator are only used for preloaded plots. In non-
  // preloaded plots min x-value is always the last seek time, and the max x-value is the current
  // playback time.
  const timeToXValueForPreloading = (t: ?Time): ?number => {
    if (xAxisVal === "timestamp" && t && startTime) {
      return toSec(subtractTimes(t, startTime));
    }
  };
  const preloadingDisplayTime = timeToXValueForPreloading(currentTime);
  const preloadingStartTime = timeToXValueForPreloading(startTime); // zero or undefined
  const preloadingEndTime = timeToXValueForPreloading(endTime);
  const hasPreloadingDisplayTime = preloadingDisplayTime != null;
  const defaultView = useMemo(() => {
    if (hasPreloadingDisplayTime) {
      if (followingViewWidth != null && parseFloat(followingViewWidth) > 0) {
        // Will be ignored in TimeBasedChart for non-preloading plots and non-timestamp plots.
        return { type: "following", width: parseFloat(followingViewWidth) };
      } else if (preloadingStartTime != null && preloadingEndTime != null) {
        return { type: "fixed", minXValue: preloadingStartTime, maxXValue: preloadingEndTime };
      }
    }
    return undefined;
  }, [followingViewWidth, hasPreloadingDisplayTime, preloadingEndTime, preloadingStartTime]);

  const onClick = useCallback((_, __, { X_AXIS_ID: seekSeconds }) => {
    if (!startTime || seekSeconds == null || !seek || xAxisVal !== "timestamp") {
      return;
    }
    // The player validates and clamps the time.
    const seekTime = TimeUtil.add(startTime, fromSec(seekSeconds));
    seek(seekTime);
  }, [seek, startTime, xAxisVal]);

  // We want to avoid rerendering the plot menu every frame, but datasets and tooltips change frequently.
  const getDatasets = useGetCurrentValue(datasets);
  const getTooltips = useGetCurrentValue(tooltips);

  const plotMenu = useMemo(
    () => (
      <PlotMenu
        displayWidth={followingViewWidth || ""}
        minYValue={minYValue}
        maxYValue={maxYValue}
        saveConfig={saveConfig}
        setMinMax={setMinMax}
        setWidth={setWidth}
        xAxisVal={xAxisVal}
        getDatasets={getDatasets}
        getTooltips={getTooltips}
      />
    ),
    [followingViewWidth, minYValue, maxYValue, saveConfig, setMinMax, setWidth, xAxisVal, getDatasets, getTooltips]
  );

  return (
    <Flex col clip center style={{ position: "relative" }}>
      <PanelToolbar helpContent={helpContent} floating menuContent={plotMenu} />
      <PlotChart
        paths={yAxisPaths}
        minYValue={parseFloat(minYValue)}
        maxYValue={parseFloat(maxYValue)}
        saveCurrentView={saveCurrentView}
        datasets={datasets}
        tooltips={tooltips}
        xAxisVal={xAxisVal}
        currentTime={preloadingDisplayTime}
        onClick={onClick}
        defaultView={defaultView}
      />
      <PlotLegend
        paths={yAxisPaths}
        saveConfig={saveConfig}
        showLegend={showLegend}
        xAxisVal={xAxisVal}
        xAxisPath={xAxisPath}
        pathsWithMismatchedDataLengths={pathsWithMismatchedDataLengths}
      />
    </Flex>
  );
}
Plot.panelType = "Plot";
Plot.defaultConfig = {
  paths: [{ value: "", enabled: true, timestampMethod: "receiveTime" }],
  minYValue: "",
  maxYValue: "",
  showLegend: true,
  xAxisVal: "timestamp",
};
Plot.shortcuts = [
  { description: "Reset", keys: ["double click"] },
  { description: "Pan", keys: ["drag"] },
  { description: "Zoom", keys: ["scroll horizontally"] },
  { description: "Zoom vertically", keys: ["v" + "scroll"] },
  { description: "Zoom both vertically and horizontally", keys: ["b" + "scroll"] },
  { description: "Zoom to percentage (10% - 100%)", keys: ["⌘", "1|2|...|9|0"] },
];

export default hot(Panel<PlotConfig>(Plot));
