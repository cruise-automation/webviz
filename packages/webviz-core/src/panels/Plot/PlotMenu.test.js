// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { flatten } from "lodash";

import { getCSVData } from "./PlotMenu";
import { getDatasetsAndTooltips } from "webviz-core/src/panels/Plot/PlotChart";

const getDatasetAndTooltipsParameters = [
  [
    {
      value: "/some_topic.ok",
      enabled: true,
      timestampMethod: "headerStamp",
    },
    {
      value: "",
      enabled: true,
      timestampMethod: "headerStamp",
    },
  ],
  {
    "/some_topic.ok": [
      [
        {
          receiveTime: {
            sec: 1570207539,
            nsec: 81366108,
          },
          headerStamp: {
            sec: 1570207538,
            nsec: 950411000,
          },
          queriedData: [
            {
              value: true,
              path: "/some_topic.ok",
            },
          ],
        },
        {
          receiveTime: {
            sec: 1570207539,
            nsec: 178513840,
          },
          headerStamp: {
            sec: 1570207539,
            nsec: 50344000,
          },
          queriedData: [
            {
              value: true,
              path: "/some_topic.ok",
            },
          ],
        },
      ],
    ],
    "": [],
  },
  {
    sec: 1570207539,
    nsec: 138873,
  },
  "timestamp",
];
const { tooltips: trackedObjectsTooltips, datasets: trackedObjectsDatasets } = getDatasetsAndTooltips(
  ...getDatasetAndTooltipsParameters
);

const tooltips = [
  {
    x: 0.010651803000000001,
    y: 0,
    datasetKey: "default",
    constantName: "",
    item: {
      receiveTime: {
        sec: 1547062466,
        nsec: 10664222,
      },
      headerStamp: {
        sec: 1547062466,
        nsec: 9726015,
      },
      queriedData: [
        {
          constantName: "",
          value: false,
          path: "/accel_vector_calibrated.using_gps_time",
        },
      ],
    },
    path: "/accel_vector_calibrated.using_gps_time",
    value: false,
    startTime: {
      sec: 1547062466,
      nsec: 12419,
    },
  },
  {
    x: 0.031882799,
    y: 0,
    datasetKey: "default",
    constantName: "",
    item: {
      receiveTime: {
        sec: 1547062466,
        nsec: 31895218,
      },
      headerStamp: {
        sec: 1547062466,
        nsec: 31719273,
      },
      queriedData: [
        {
          constantName: "",
          value: false,
          path: "/accel_vector_calibrated.using_gps_time",
        },
      ],
    },
    path: "/accel_vector_calibrated.using_gps_time",
    value: false,
    startTime: {
      sec: 1547062466,
      nsec: 12419,
    },
  },
];
const data = tooltips.map(({ x, y }) => ({ x, y }));

const datasetsSingleTopic = [
  {
    borderColor: "#4e98e2",
    label: "/accel_vector_calibrated.using_gps_time",
    key: "default",
    showLine: true,
    fill: false,
    borderWidth: 1,
    pointRadius: 1.5,
    pointHoverRadius: 3,
    pointBackgroundColor: "#74beff",
    pointBorderColor: "transparent",
    data,
  },
];

const datasetsMultipleTopics = [1, 2, 3].map((num) => ({
  borderColor: "#4e98e2",
  label: `/topic${num}`,
  key: `dataset-${num}`,
  showLine: true,
  fill: false,
  borderWidth: 1,
  pointRadius: 1.5,
  pointHoverRadius: 3,
  pointBackgroundColor: "#74beff",
  pointBorderColor: "transparent",
  data: data.map(({ x, y }) => ({ x: x + num, y: y + num })),
}));

const tooltipsMultipleTopics = flatten(
  [1, 2, 3].map((num) =>
    tooltips.map((tooltip) => ({
      ...tooltip,
      datasetKey: `dataset-${num}`,
    }))
  )
);

const tooltipsNoHeader = flatten(
  [1, 2, 3].map((num) =>
    tooltips.map((tooltip) => ({
      ...tooltip,
      datasetKey: `dataset-${num}`,
      item: {
        ...tooltip.item,
        headerStamp: num === 3 ? undefined : tooltip.item.headerStamp,
      },
    }))
  )
);

const tooltipsDiffTimestamp = flatten(
  [1, 2, 3].map((num) =>
    tooltips.map((tooltip) => ({
      ...tooltip,
      datasetKey: `dataset-${num}`,
      item: {
        ...tooltip.item,
        headerStamp: tooltip.item.headerStamp && {
          ...tooltip.item.headerStamp,
          sec: tooltip.item.headerStamp.sec + num,
        },
      },
    }))
  )
);

describe("PlotMenu", () => {
  it("Single topic", () => {
    expect(getCSVData(datasetsSingleTopic, tooltips, "timestamp")).toMatchSnapshot();
  });

  it("Multiple topics", () => {
    expect(getCSVData(datasetsMultipleTopics, tooltipsMultipleTopics, "timestamp")).toMatchSnapshot();
  });

  it("Multiple topics with one topic don't have header.stamp", () => {
    expect(getCSVData(datasetsMultipleTopics, tooltipsNoHeader, "timestamp")).toMatchSnapshot();
  });

  it("Multiple topics with different header.stamp", () => {
    expect(getCSVData(datasetsMultipleTopics, tooltipsDiffTimestamp, "timestamp")).toMatchSnapshot();
  });

  it("works with data directly from getDatasetsAndTooltips", () => {
    expect(getCSVData(trackedObjectsDatasets, trackedObjectsTooltips, "timestamp")).toMatchSnapshot();
  });
});
