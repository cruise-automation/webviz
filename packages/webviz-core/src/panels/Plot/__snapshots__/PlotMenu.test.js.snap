// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlotMenu Multiple topics 1`] = `
"elapsed time,receive time,header.stamp,topic,value
1.010651803,1547062466.010664222,1547062466.009726015,/topic1,1
1.031882799,1547062466.010664222,1547062466.009726015,/topic1,1
2.010651803,1547062466.010664222,1547062466.009726015,/topic2,2
2.031882799,1547062466.010664222,1547062466.009726015,/topic2,2
3.010651803,1547062466.010664222,1547062466.009726015,/topic3,3
3.031882799,1547062466.010664222,1547062466.009726015,/topic3,3"
`;

exports[`PlotMenu Multiple topics with different header.stamp 1`] = `
"elapsed time,receive time,header.stamp,topic,value
1.010651803,1547062466.010664222,1547062467.009726015,/topic1,1
1.031882799,1547062466.010664222,1547062467.009726015,/topic1,1
2.010651803,1547062466.010664222,1547062468.009726015,/topic2,2
2.031882799,1547062466.010664222,1547062468.009726015,/topic2,2
3.010651803,1547062466.010664222,1547062469.009726015,/topic3,3
3.031882799,1547062466.010664222,1547062469.009726015,/topic3,3"
`;

exports[`PlotMenu Multiple topics with one topic don't have header.stamp 1`] = `
"elapsed time,receive time,header.stamp,topic,value
1.010651803,1547062466.010664222,1547062466.009726015,/topic1,1
1.031882799,1547062466.010664222,1547062466.009726015,/topic1,1
2.010651803,1547062466.010664222,1547062466.009726015,/topic2,2
2.031882799,1547062466.010664222,1547062466.009726015,/topic2,2
3.010651803,1547062466.010664222,,/topic3,3
3.031882799,1547062466.010664222,,/topic3,3"
`;

exports[`PlotMenu Single topic 1`] = `
"elapsed time,receive time,header.stamp,topic,value
0.010651803000000001,1547062466.010664222,1547062466.009726015,/accel_vector_calibrated.using_gps_time,0
0.031882799,1547062466.010664222,1547062466.009726015,/accel_vector_calibrated.using_gps_time,0"
`;

exports[`PlotMenu works with data directly from getDatasetsAndTooltips 1`] = `
"elapsed time,receive time,header.stamp,topic,value
-0.04972787299999992,1570207539.081366108,1570207538.950411000,/some_topic.ok,1
0.050205127,1570207539.081366108,1570207538.950411000,/some_topic.ok,1"
`;
